### ISSUE TITLE
Help with GitHub Enterprise

### ISSUE THREAD

**Issue by mdwheele** [mdwheele/classroom#1]
Related: https://github.com/education/classroom/issues/1213

@trombonekenny would you mind sending me the steps you're using to reproduce, pretty much starting from scratch? I know that's a lot to ask, but I'd like to factor differences in deployment of Enterprise out of the equation so more information is better, for me. Currently, I cannot reproduce this specific issue but I'm down to help. 

I have an open PR (https://github.com/education/classroom/pull/1163) to support GitHub Enterprise so if there is an actual bug, I want to add tests to reproduce. Currently, tests are passing.

**Comment by trombonekenny** [comment]
Yeah, I'll see if I can get the steps down to reproduce in a docker environment and I'll send that over.

**Comment by mdwheele** [comment]
Send your Dockerfile/compose.yml as well if you can. Would be helpful.

**Comment by trombonekenny** [comment]
Docker compose... that would be an elegant way to do it. :) I was just ham handing it into a single `centos` container because that's what I know. But I was able to get an instance of Classroom dev working from inside a container (with only minor changes and acceptable swearing and bruising). The problem was still there, so maybe this is worth hunting down for the PR.

But first the easy data, Enterprise version 2.9.10. Running your `classroom:feature/support-ghe` branch.

Error is the same as referenced in https://github.com/education/classroom/issues/1213

More details on what I did to follow.

**Comment by trombonekenny** [comment]
Here's roughly what I did. It's not a shell script to run and might not be complete (so it's not like a cooked Dockerfile) but hopefully it makes sense. Even if some of the details are missing, the top to bottom order is consistent.

docker run -it --name classroom -p 5000:5000 -v ~/classroom:/classroom centos /bin/bash

yum install -y postgresql-server git vim bzip2 gcc make \\
openssl-devel readline-devel zlib-devel postgresql-devel gcc-c++ \\
epel-release java-1.8.0-openjdk memcached

curl https://dl.yarnpkg.com/rpm/yarn.repo -o /etc/yum.repos.d/yarn.repo
curl --silent --location https://rpm.nodesource.com/setup_6.x | bash -

yum install -y yarn redis

wget https://download.elastic.co/elasticsearch/elasticsearch/elasticsearch-1.7.3.noarch.rpm 
yum localinstall elasticsearch-1.7.3.noarch.rpm

git clone https://github.com/rbenv/rbenv.git ~/.rbenv

echo 'export PATH="~/.rbenv/bin:$PATH"' >> ~/.bash_profile
echo 'eval "$(rbenv init -)"' >> ~/.bash_profile

bash -l   # to get root to read the .bash_profile

mkdir -p "$(rbenv root)"/plugins
git clone https://github.com/rbenv/ruby-build.git "$(rbenv root)"/plugins/ruby-build

rbenv install 2.4.2
rbenv global 2.4.2

su postgres -c 'initdb -E UTF8 -D /var/lib/pgsql/data'
/var/lib/psgql/data/postgresql.conf  uncomment #unix_socket_permissions = 0777
su postgres -c 'pg_ctl -D /var/lib/pgsql/data start'
psql    
CREATE USER root WITH SUPERUSER;

gem install bundler && rbenv rehash

git clone https://github.com/mdwheele/classroom.git
cd classroom
git checkout feature/support-ghe

script/setup

Edited .env changing these values
ENTERPRISE_ENABLED
GITHUB_ENTERPRISE_HOSTNAME=

GITHUB_CLIENT_ID
GITHUB_CLIENT_SECRET
CLASSROOM_WEBHOOK_URL_PREFIX
WEBHOOK_SECRET

To get Classroom running as root in the docker environment, I changed Procfile.dev to look like this:
```
# Start development mode executables.
# This assumes the postgresql is already running
elasticsearch: /usr/share/elasticsearch/bin/elasticsearch -Des.default.path.conf=/etc/elasticsearch
memcached:     memcached -u memcached
redis:         redis-server

web: ./bin/rails server -p 5000 -b 0.0.0.0
worker: ./bin/sidekiq -C config/sidekiq.yml
```
script/server

**Comment by mdwheele** [comment]
Thanks very much for the detail! I can find some time tomorrow to take a deeper look.

**Comment by trombonekenny** [comment]
Any updates? Speaking of that, I'll make sure my GHE is updated to the latest patches. But I am curious if you're on a different point release than I am -- currently running 2.9.

**Comment by mdwheele** [comment]
Yeah sorry about that @trombonekenny. I meant to respond I was delayed. I'll get to this as soon as I can. We're running 2.11.2 right now I think so there *may* be differences there but I really doubt it. FWIW, I did get a chance to try and reproduce using our setup and can't so maaaaybe the problem is there but I haven't had time to go through your steps. 

I'll try and make time this week. We're just in the middle of doing some work with OpenShift so it's been difficult to get time.

**Comment by mdwheele** [comment]
Hrmmmm, I got some time tonight to attempt to reproduce using the steps you provide and it works for me. Before we proceed, I'd like to take some time in the morning to do a little more diagnostics and see if I can trace what might be happening with invites (i.e. find out reasons why `github_repository#invite` could return `nil`). I *have* had `script/setup` go sideways and just not create tables or mess other things up. The initial setup for this project seems so brittle. 

I'll poke here again tomorrow. I've also got a `docker-compose`-based repository that I have been using for local development that pulls images for elasticsearch, redis, postgres, memcached and creates a development workspace container for running Rails commands. Lemme know if you're interested and I can share.

**Comment by trombonekenny** [comment]
Hey great! I would post output of `script/setup` but the terminals I ran it in are long closed. I'm not sure how useful that is on subsequent executions.

I started tooling with docker-compose and it does seem like a much smarter way to do this development. If you share your compose config, I can try to duplicate it there. Thanks!

**Comment by mdwheele** [comment]
Gonna paste a bunch here regarding how my run went:

![image](https://user-images.githubusercontent.com/2453394/33391330-a3d31248-d506-11e7-9779-32a3f9408bc2.png)

We have Classroom! I click `Sign in`:

![image](https://user-images.githubusercontent.com/2453394/33391355-b9ce0ab2-d506-11e7-8986-b111a694ae77.png)

A familiar face... I've previously approved this OAuth application, so it sends me straight to GitHub Classroom:

![image](https://user-images.githubusercontent.com/2453394/33391387-e02482f4-d506-11e7-8a39-e3cb8427a1f9.png)

I click through a few screens to set up the first classroom. I create a new individual assignment:

![image](https://user-images.githubusercontent.com/2453394/33391470-22b2d35a-d507-11e7-91cb-e482e7e1a97c.png)

I copy the invitation link and visit in a new tab:

![image](https://user-images.githubusercontent.com/2453394/33391499-3ee63ee0-d507-11e7-9e9f-29002568ffb9.png)

HA! I can reproduce the issue now!

![image](https://user-images.githubusercontent.com/2453394/33391613-999274bc-d507-11e7-96d9-4f8084361ce3.png)

I'm going to rebase on `upstream/master` and see if the problem persists. This part of the application wasn't touched in my PR and tests passed at time of submission. Things might have changed in the meantime. Otherwise, I'll be digging into it further. Super stoked I can reproduce it though.
